aws/resource_aws_pinpoint_app.go

```diff
@@ -0,0 +1,435 @@
+	} else {
+		name = resource.UniqueId()
+	}
+
+	log.Printf("[DEBUG] Pinpoint create app: %s", name)
+
+	req := &pinpoint.CreateAppInput{
+		CreateApplicationRequest: &pinpoint.CreateApplicationRequest{
+			Name: aws.String(name),
+		},
+	}
+
+	output, err := pinpointconn.CreateApp(req)
+	if err != nil {
+		return fmt.Errorf("[ERROR] creating Pinpoint app: %s", err)
+	}
+
+	d.SetId(*output.ApplicationResponse.Id)
+
+	return resourceAwsPinpointAppUpdate(d, meta)
+}
+
+func resourceAwsPinpointAppUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	appSettings := &pinpoint.WriteApplicationSettingsRequest{}
+
+	//if d.HasChange("cloudwatch_metrics_enabled") {
+	//	appSettings.CloudWatchMetricsEnabled = aws.Bool(d.Get("cloudwatch_metrics_enabled").(bool));
+	//}
+
+	if d.HasChange("campaign_hook") {
+		appSettings.CampaignHook = expandCampaignHook(d)
+	}
+
+	if d.HasChange("limits") {
+		appSettings.Limits = expandCampaignLimits(d.Get("limits").(*schema.Set))
+	}
+
+	if d.HasChange("quiet_time") {
+		appSettings.QuietTime = expandQuietTime(d.Get("quiet_time").(*schema.Set))
+	}
+
+	req := pinpoint.UpdateApplicationSettingsInput{
+		ApplicationId:                   aws.String(d.Id()),
+		WriteApplicationSettingsRequest: appSettings,
+	}
+
+	_, err := conn.UpdateApplicationSettings(&req)
+	if err != nil {
+		return err
+	}
+
+	return resourceAwsPinpointAppRead(d, meta)
+}
+
+func resourceAwsPinpointAppRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[INFO] Reading Pinpoint App Attributes for %s", d.Id())
+
+	app, err := conn.GetApp(&pinpoint.GetAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	settings, err := conn.GetApplicationSettings(&pinpoint.GetApplicationSettingsInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
+		if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
+			log.Printf("[WARN] Pinpoint App (%s) not found, error code (404)", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return err
+	}
+
+	d.Set("name", app.ApplicationResponse.Name)
+	d.Set("application_id", app.ApplicationResponse.Id)
+
+	d.Set("campaign_hook", flattenCampaignHook(settings.ApplicationSettingsResource.CampaignHook))
+	d.Set("limits", flattenCampaignLimits(settings.ApplicationSettingsResource.Limits))
+	d.Set("quiet_time", flattenQuietTime(settings.ApplicationSettingsResource.QuietTime))
+
+	return nil
+}
+
+func resourceAwsPinpointAppDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).pinpointconn
+
+	log.Printf("[DEBUG] Pinpoint Delete App: %s", d.Id())
+	_, err := conn.DeleteApp(&pinpoint.DeleteAppInput{
+		ApplicationId: aws.String(d.Id()),
+	})
+	if err != nil {
```

Comment:
We should add the following above this to ignore if its already been deleted:

```go
if isAWSErr(err, pinpoint.ErrCodeNotFoundException, "") {
  return nil
}
```

Reply:
Fixed in https://github.com/terraform-providers/terraform-provider-aws/pull/5956/commits/be7dc0047e5e3ec5d966f2a51c702546a27cfb95